components/experimental/src/dimension/currency/long_format.rs

```diff
@@ -0,0 +1,127 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::str::FromStr;
+
+use fixed_decimal::FixedDecimal;
+
+use icu_decimal::FixedDecimalFormatter;
+use icu_pattern::DoublePlaceholderPattern;
+use icu_plurals::PluralRules;
+use writeable::Writeable;
+
+use crate::dimension::provider::currency_patterns::{CurrencyPatternsDataV1, PatternCount};
+use crate::dimension::provider::extended_currency::{Count, CurrencyExtendedDataV1};
+
+use super::CurrencyCode;
+
+pub struct LongFormattedCurrency<'l> {
+    pub(crate) value: &'l FixedDecimal,
+    pub(crate) currency_code: CurrencyCode,
+    pub(crate) extended: &'l CurrencyExtendedDataV1<'l>,
+    pub(crate) patterns: &'l CurrencyPatternsDataV1<'l>,
+    pub(crate) fixed_decimal_formatter: &'l FixedDecimalFormatter,
+    pub(crate) plural_rules: &'l PluralRules,
+}
+
+writeable::impl_display_with_writeable!(LongFormattedCurrency<'_>);
+
+impl<'l> Writeable for LongFormattedCurrency<'l> {
+    fn write_to<W>(&self, sink: &mut W) -> core::result::Result<(), core::fmt::Error>
+    where
+        W: core::fmt::Write + ?Sized,
+    {
+        let plural_category = self.plural_rules.category_for(self.value);
+        let count = Count::from(plural_category);
+        let pattern_count = PatternCount::from(plural_category);
+
+        let display_names = &self.extended.display_names;
+
+        let display_name = display_names
+            .get(&count)
+            .or_else(|| display_names.get(&Count::Other))
+            // TOOD: check the cases of using `Count::DisplayName`
+            .unwrap_or_else(|| self.currency_code.0.as_str());
+
+        let patterns = &self.patterns.unit_patterns;
+        let pattern = patterns
+            .get(&pattern_count)
+            .or_else(|| patterns.get(&PatternCount::Other))
+            .ok_or(core::fmt::Error)?;
+
+        // Parse the pattern string into a DoublePlaceholderPattern
+        // The pattern is expected to be in the form of "{0} {1}"
+        let pattern = DoublePlaceholderPattern::from_str(pattern).map_err(|_| core::fmt::Error)?;
```

Comment:
I don't like adding this runtime parsing code, you should really change the data structs to use `Pattern` soon.

Reply:
yes, we need that of course.

Reply:
done